Use b3dm tilesets for classification

bagnell opened the pull request:
This adds a `classificationType` option to `Cesium3DTileset` that will turn the glTF of a b3dm into a classification volume.

`ClassificationModel` has support for some of the same options that `Model` does (e.g. terrain clamping). Can this be removed or should we make this part of the public API?

Limitations:
* only works for b3dm
* all indices with the same batch id must be contiguous in memory
* ignores all shaders and techniques
* the only extension supported is `WEB3D_quantized_attributes`

Note that this is a PR into the vector-tiles branch.

cesium-concierge commented:
![Signed CLA is on file.](https://img.shields.io/badge/CLA-signed-brightgreen.svg)

@bagnell, thanks for the pull request! Maintainers, we have a signed CLA from @bagnell, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

pjcozzi commented:
Limitations all sound OK to me.  Later I could see users wanting to classify with i3dm - it is is kind of in between `b3dm` and `geom`.

pjcozzi commented:
> ClassificationModel has support for some of the same options that Model does (e.g. terrain clamping). Can this be removed or should we make this part of the public API?

I don't think `ClassificationModel` needs to be exposed through the public API; Cesium API users will use the `ClassificationPrimitive` if they are not using 3D Tiles...I think.

So it only needs to have the API needed to support `Cesium3DTileset`, which does not have terrain clamping I believe.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1053,6 +1046,19 @@ define([
+        },
+
+        /**
+         * Determines whether terrain, 3D Tiles or both will be classified by this tileset.
```
Label as `@readonly`.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -222,6 +222,8 @@ define([
 
         this._readyPromise = when.defer();
 
+        this._classificationType = options.classificationType;
```
Is this in the reference doc?

pjcozzi replied:
Also include the limitations in the reference doc.  It might be easiest to add this once in the property getter and then reference it from the constructor function.

pjcozzi commented:
Update CHANGES.md.

pjcozzi commented:
Add a Sandcastle example.

pjcozzi commented on Source/Scene/Vector3DTilePrimitive.js:
```diff
@@ -153,8 +153,18 @@ define([
          */
-        this.classificationType = ClassificationType.CESIUM_3D_TILE;
-        this._classificationType = this.classificationType;
+        this.classificationType = defaultValue(options.classificationType, ClassificationType.CESIUM_3D_TILE);
```
Add to reference doc even though the class is private.

pjcozzi commented:
Add unit tests.

pjcozzi commented on Source/Scene/Batched3DModel3DTileContent.js:
```diff
@@ -354,29 +377,50 @@ define([
+                pickObject : pickObject
+            });
+        } else {
+            content._model = new ClassificationModel({
```
Can you add a code comment about how this is a transcode - with limitations on the input glTF - so that we can use the `geom` rebatching rendering pipeline?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,2482 @@
+define([
```
Looks like it does more than it needs to given the limitations classification has.  Are you sure there aren't other limitations we can add to make this even more simple, e.g., vertex formats, removing the cache (is it needed), etc.

Then perhaps factor out some common functions to share with `Model`?

bagnell commented:
@pjcozzi This is ready for another look.

pjcozzi commented:
Do you get test failures like this:

![image](https://user-images.githubusercontent.com/782098/33797652-e3a6f2b0-dcd9-11e7-8d70-7a98bbf98899.png)

when running http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/b3dm-class/Specs/SpecRunner.html

I get a total of

> 9590 specs, 96 failures, 6 pending specs

Some may just be before I am running from S3.

MacBook Pro / AMD.

pjcozzi commented on Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html:
```diff
@@ -0,0 +1,66 @@
+var classification = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
+    url: 'https://beta.cesium.com/api/assets/3486?access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiIzZjM4MjljZi0xNTAzLTQ0MzctODhlZi0zOTU3Njg5ODA1Y2QiLCJpZCI6OCwiaWF0IjoxNDgxODI5ODMyfQ.dYlV8_PoXcFNlPHGook5kMzuJMS-Bb9DCMzI1mFVvgE',
+    classificationType : Cesium.ClassificationType.CESIUM_3D_TILE,
+    skipLevelOfDetail : false
```
Can this be omitted?

I thought the plan was to update the documentation for `skipLevelOfDetail` so that it was a "hint" - and then to just implicitly disable it when a tileset is used for classification.

pjcozzi commented on Apps/Sandcastle/gallery/3D Tiles Photogrammetry Classification.html:
```diff
@@ -0,0 +1,66 @@
+    throw(error);
+});
+
+var classification = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
```
What does this tileset look like without classification.  Is it worth adding a comment to say to comment out the `classificationType` line to see the geometry used to classify?

The other option is to create two tilesets, one with and without the classification, and then change their `show` via a checkbox, but that could make the example too complex.

pjcozzi commented on CHANGES.md:
```diff
@@ -9,6 +9,13 @@ Change Log
+* Added `pack` and `unpack` functions to `OrientedBoundingBox` for packing to and unpacking from a flat buffer.
+* Added experimental support for [3D Tiles Vector Data](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/vector-tiles/TileFormats/VectorData) ([#4665](https://github.com/AnalyticalGraphicsInc/cesium/pull/4665)). The new and modified Cesium APIs are:
+   * `Cesium3DTileStyle` has expanded for styling point features. See the [styling specification](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/vector-tiles/Styling#vector-data) for details.
+   * `Cesium3DTileFeature` can modify `color` and `show` properties for polygon, polyline, geometry, and mesh features.
```
Is this still right?  I thought `mesh` features were removed.

bagnell replied:
They were removed in the other PR.

pjcozzi commented on Source/Shaders/Builtin/Functions/writeDepthClampedToFarPlane.glsl:
```diff
@@ -0,0 +1,24 @@
+ * gl_FragColor = color;
+ * czm_writeDepthClampedToFarPlane();
+ *
+ * @see czm_writeDepthClampedToFarPlane
```
This `@see` references the same function.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+     *
+     * @demo {@link http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=3D%20Models.html|Cesium Sandcastle Models Demo}
+     */
+    function ClassificationModel(options) {
```
This still needs tests, right?  We should test this as an isolated unit, and not just rely on the 3D Tiles tests.

bagnell replied:
I started writing tests here but it required duplicating a ton of code from `Batched3DModel3DTileContent` in the spec.

pjcozzi replied:
Is it possible that when you create a common file for both `Model` implementations that it will because easier to write unit tests since you can also write tests against that?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+        './Vector3DTileBatch',
+        './Vector3DTilePrimitive'
+    ], function(
+        arraySlice,
```
Did you remove unused defines after the latest refactor?

bagnell replied:
Removed.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+    var ModelState = {
+        NEEDS_LOAD : 0,
+        LOADING : 1,
+        LOADED : 2,  // Renderable, but textures can still be pending when incrementallyLoadTextures is true.
```
Comment doesn't matter in this context, right?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+            gltf = new Uint8Array(gltf);
+        }
+
+        if (gltf instanceof Uint8Array) {
```
Should this only support Binary glTF?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+        this.debugWireframe = defaultValue(options.debugWireframe, false);
+        this._debugWireframe = false;
+
+        this._distanceDisplayCondition = options.distanceDisplayCondition;
```
Remind me where this is used via 3D Tiles?

bagnell replied:
I thought it was used, but was not. Removed.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+            scale : undefined,
+
+            // Per-node show inherited from parent
+            computedShow : true,
```
Here and below, does this still need to be simplified give that we will only support one mesh?  Did we confirm that only one primitive is OK as well?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+         *
+         * @private
+         */
+        texturesByteLength : {
```
Just return zero?

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1869 @@
+        } else {
+            // TRS converted to Cesium types
+            var rotation = gltfNode.rotation;
+            runtimeNode.translation = Cartesian3.from
```
Seems like once this is trimmed down for one node/primitive, there are some obvious things that can be factored into a common file such as the quantization functions.  I wonder if extension handling can also be made more simple here since only quantization and RTC need to be supported.

Also remind me, can we completely ignore the shaders since we are assuming no alpha texture / discard?

bagnell replied:
RTC, binary glTF, and quantization are the extensions that are supported. The other glTF extensions are for the material and technique which we don't even parse.

We now also completely ignore the shaders.

bagnell commented:
> Do you get test failures like this:

I don't see any test failure locally and the Travis tests pass. Also, running the tests from the link you posted give strange failures like:

Core/Cartesian3 renders with external gltf

bagnell commented:
The requirement for floating point positions and uint16 batch ids were removed as well.Different formats are required to support quantization.

bagnell commented:
@pjcozzi I think this is ready. The only thing missing would be only supporting one primitive per mesh, but I could do that in a separate PR.

pjcozzi commented:
> The only thing missing would be only supporting one primitive per mesh, but I could do that in a separate PR.

Separate PR is OK, but let's do it right away to keep the code streamlined.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1807 @@
+        };
+    }
+
+    function computeBoundingSphere(model) {
```
Are you going to move some of these functions to a common file?  Or do you want to do that in a separate PR?

bagnell commented:
@pjcozzi This is ready for another look. I added the restriction that only one primitive per mesh is supported, factored out some of the common code with Model, and added ClassificationModel tests.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1209 @@
+    var aMinScratch = new Cartesian3();
+    var aMaxScratch = new Cartesian3();
+
+    function computeBoundingSphere(model) {
```
Can this be a common function?

bagnell replied:
This function is different enough from model that I didn't create a common function. The only part I could pull out is finding the min and max.

pjcozzi commented on Source/Scene/ClassificationModel.js:
```diff
@@ -0,0 +1,1209 @@
+        }
+    }
+
+    function createCommand(model) {
```
Maybe rename this since it no longer creates a command...directly.

pjcozzi commented:
Tests and Sandcastle example look good.  Just those trivial questions.

bagnell commented:
@pjcozzi This should be ready.

pjcozzi commented:
Thanks!
